tracer/src/Datadog.Trace/Baggage.cs

```diff
@@ -0,0 +1,354 @@
+
+    private static readonly List<KeyValuePair<string, string>> EmptyList = [];
+
+    private List<KeyValuePair<string, string>>? _items;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="Baggage"/> class.
+    /// </summary>
+    public Baggage()
+    {
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="Baggage"/> class using the specified items.
+    /// </summary>
+    /// <param name="items">The baggage items.</param>
+    public Baggage(IEnumerable<KeyValuePair<string, string>> items)
+    {
+        if (items != null!)
+        {
+            _items = [..items];
+        }
+    }
+
+    /// <summary>
+    /// Gets or sets the baggage collection for the current execution context.
+    /// </summary>
+    public static Baggage Current
+    {
+        get => AsyncStorage.Value ??= new Baggage();
+        set => AsyncStorage.Value = value;
+    }
+
+    /// <summary>
+    /// Gets the count of items in this baggage instance.
+    /// </summary>
+    public int Count => _items?.Count ?? 0;
+
+    bool ICollection<KeyValuePair<string, string>>.IsReadOnly => false;
+
+    ICollection<string> IDictionary<string, string>.Keys
+    {
+        get
+        {
+            if (_items is { Count: > 0 } items)
+            {
+                var keys = new string[items.Count];
+
+                for (int i = 0; i < items.Count; i++)
+                {
+                    keys[i] = items[i].Key;
+                }
+
+                return keys;
+            }
+
+            return [];
+        }
+    }
+
+    ICollection<string> IDictionary<string, string>.Values
+    {
+        get
+        {
+            if (_items is { Count: > 0 } items)
+            {
+                var values = new string[items.Count];
+
+                for (int i = 0; i < items.Count; i++)
+                {
+                    values[i] = items[i].Value;
+                }
+
+                return values;
+            }
+
+            return [];
+        }
+    }
+
+    public string this[string key]
+    {
+        get
+        {
+            if (TryGetValue(key, out var value))
+            {
+                return value;
+            }
+
+            ThrowHelper.ThrowKeyNotFoundException($"The key was not found: {key}");
+            return default!; // unreachable
+        }
+
+        set
+        {
+            AddOrReplace(key, value);
+        }
+    }
+
+    private List<KeyValuePair<string, string>> EnsureListInitialized()
+    {
+        if (_items == null)
+        {
+            Interlocked.CompareExchange(ref _items, [], null);
+        }
+
+        return _items;
+    }
+
+    bool ICollection<KeyValuePair<string, string>>.Remove(KeyValuePair<string, string> item)
+    {
+        if (item.Key == null!)
+        {
+            ThrowHelper.ThrowArgumentException("The key cannot be null.", nameof(item));
+        }
+
+        if (_items is { Count: > 0 } list)
+        {
+            lock (list)
```

Comment:
Is locking on `_items` (`list` here I guess) okay? I always thought you wanted to lock on some dedicated `readonly object` that was specific for locking?

Reply:
It should be ok here since
- the lock is used to control access to the list itself, nothing else
- if there is no list, we don't need a lock
- the list is private and never shared outside of `Baggage`, so there's no danger of some other code also locking on the same object and risking a deadlock

An implementation with a separate `object` for locking _could_ be used. Such an object could also guard when initializing `_items`, but this way we can avoid allocating that extra `object`. Here I am handling the initialization of `_items` using `Interlocked.CompareExchange()` in `Baggage.EnsureListInitialized()`, no separate lock required.

Reply:
> the list is private and never shared outside of `Baggage`, so there's no danger of some other code also locking on the same object and risking a deadlock

@tonyredondo did find a case in `MergeInto()` where we could deadlock because we lock both the list from `this` and from another baggage instance. Fixed.